content/blog/fp-js-higher-order-functions.md

```diff
@@ -0,0 +1,520 @@
+
+And so on.
+
+So, beware of this when using `map`, `filter`, `reduce` and passing the callback function directly like above when it's type signature correspons with the callback signature.
+
+## Other notable examples
+
+There are many other higher order functions built in right to the language. We won't dive any deeper into them, but here is a list of some that you probably will or already are using on a regular basis.
+
+```js
+setTimeout(() => console.log('I run after a second!'), 1000);
+setInterval(() => console.log('I run every second!'), 1000);
+element.addEventListener('click', () => console.log('I run on clicks!'));
+```
+
+## Writing our own higher order function!
+
+Now that you are somewhat familiar with higher order functions, both those that can return functions and those that take functions as arguments, lets try to build some of our own, useful higher order functions!
+
+### Drying up code with higher order functions
+
+Lets say you are building a parser for something. This parser needs to deal with lists of generic elements. These elements all share some properties in common, but can differ alot. Lets also say that a property that they all share is `type`, that tells us what type of element this is. Before we can operate on an element, we need to assert the type of it, otherwise our parser will blow up!
+
+```js
+const elements = [
+  {type: 'a', href: '/checkout'},
+  {type: 'div'},
+  {type: 'p'},
+  {type: 'span'},
+  {type: 'blink', isBlinking: false},
+  {type: 'marquee'},
+  {type: 'cthulu', isSleeping: true},
+  {type: 'a', href: '/contact'}
+];
+```
+
+So, we got our list of elements that we need to parse, now we have to build some functions that can assert what type they are, so that we do not try to access properties that do not exist. One implementation of this could look like this:
+
+```js
+function isTypeA(element) {
+  return element.type === 'a';
+}
+
+function isTypeDiv(element) {
+  return element.type === 'div';
+}
+
+function isTypeCthulu(element) {
+  return element.type === 'cthulu';
+}
+// And so on
+```
+
+Seems like we are repeating ourselves pretty much, almost all the functions look exactly the same. Definetly room for improvment.
+
+Now, remember the `add` example, were we made multiple "adders" with one function that returned another? Lets apply the same principle here.
+
+First, we make a higher order function, one that returns a new function. We want the returned function to assert that the element is of a certain type.
+
+```js
+function makeTypeChecker() {
+  return function checkType(element) {
+    return element.type === // what exactly?
+  }
+}
+```
+
+Next step is to figure out what to check the element.type against. Remember, that the inner function can remember arguments and variables from the outer function due to scope. So, lets pass in a string to check the type against.
```

Comment:
**Suggestion**
Remember that the outer and inner function result in a closure and the inner function can remember arguments and variables from the outer function due to the outer function's containing scope.

Reply:
Fixed.